applications/hpcc/hpcc_random_access.cpp

```diff
@@ -0,0 +1,120 @@
+// to run, do something like
+// make -j demo-hpcc_random_access
+// bin/grappa_run --ppn 8 --nnode 12 -- demo-hpcc_random_access.exe 
+
+#include <Grappa.hpp>
+#include <iomanip>
+
+using namespace Grappa;
+
+// define command-line flags (third-party 'gflags' library)
+DEFINE_int64( scale, 20, "table size = 2 ^ scale * sizeof(uint64_t) * number of PEs" );
+DEFINE_int64( iters, 4, "Number of iterations" );
+
+// define custom statistics which are logged by the runtime
+// (here we're not using these features, just printing them ourselves)
+GRAPPA_DEFINE_METRIC( SimpleMetric<double>, gups_runtime, 0.0 );
+GRAPPA_DEFINE_METRIC( SimpleMetric<double>, gups_throughput, 0.0 );
+
+#define POLY 0x0000000000000007ULL
+#define PERIOD 1317624576693539401LL
+
+typedef double time_type;
+std::string print_time(time_type t)
+{
+	std::ostringstream out;
+	out << std::setiosflags(std::ios::fixed) << std::setprecision(2) << t;
+	return out.str();
+}
+
+
+uint64_t N;
+uint64_t HPCC_starts(int64_t n);
+
+Grappa::GlobalCompletionEvent randomaccess_gce;
+
+template <Grappa::GlobalCompletionEvent * GCE = &randomaccess_gce >
+	double run_random_access() {
+		LOG(INFO) << "HPCC RandomAccess" << std::endl;
+		N = (1 << FLAGS_scale) * cores();
+
+		// create target array that we'll be updating
+		auto hpcc_table = global_alloc<int64_t>(N);
+		Grappa::memset( hpcc_table, 0, N);
+
+		double tstart = walltime();
+
+		on_all_cores([hpcc_table] {
+			//for(int k = 0; k < FLAGS_iters; k++)
+			//forall_here<async>(0, FLAGS_iters, [=](uint64_t k) {
+			forall_here<Grappa::SyncMode::Async>(0, FLAGS_iters, [=](int64_t k) {
```

Comment:
`forall_here<Async>` actually explicitly does *not* wait for its children. you should be able to stick with your normal `for` loop from before here...